onnxruntime/core/providers/cpu/tensor/cast_op.h

```diff
@@ -66,6 +66,82 @@ inline void CastFloat16Data(const Tensor* in, Tensor* out, const TensorShape& sh
   allocator->Free(buffer);
 }
 
+template <typename SrcType>
+inline void CastToStringData(const Tensor* in, Tensor* out, const TensorShape& shape) {
+  const int64_t len = shape.Size();
+  ORT_ENFORCE(len > 0);
+  for (int i = 0; i < len; ++i) {
+    if (std::is_floating_point<SrcType>::value && std::isnan(in->Data<SrcType>()[i])) {
+      out->MutableData<std::string>()[i] = "NaN";
+    } else if (std::is_floating_point<SrcType>::value && std::isinf(in->Data<SrcType>()[i])) {
+      if (in->Data<SrcType>()[i] < std::numeric_limits<SrcType>::lowest()) {
+        out->MutableData<std::string>()[i] = "-INF";
+      } else {
+        out->MutableData<std::string>()[i] = "INF";
+      }
+    } else {
+      std::ostringstream convert;
+      convert.precision(20);
+      convert << in->Data<SrcType>()[i];
+      out->MutableData<std::string>()[i] = convert.str();
+    }
+  }
+}
+
+template <typename DstType>
+inline void CastFromStringData(const Tensor* in, Tensor* out, const TensorShape& shape) {
+  if (std::is_same<DstType, std::string>::value) return;
+  const int64_t len = shape.Size();
+  ORT_ENFORCE(len > 0);
+  if (std::is_same<DstType, float>::value) {
+    for (int i = 0; i < len; ++i) {
+      out->MutableData<float>()[i] = std::stof(in->Data<std::string>()[i]);
+    }
```

Comment:
Can we move the calls to MutableData and Data outside of the loop unless you're sure they'll get optimized to pull them up and not incur the function call cost on each iteration